src/rgw/rgw_op.cc

```diff
@@ -5276,156 +5276,184 @@ void RGWCompleteMultipart::execute()
+  
+    etag_bl.append(final_etag_str, strlen(final_etag_str) + 1);
+  
+    attrs[RGW_ATTR_ETAG] = etag_bl;
+  
+    if (compressed) {
+      // write compression attribute to full object
+      bufferlist tmp;
+      ::encode(cs_info, tmp);
+      attrs[RGW_ATTR_COMPRESSION] = tmp;
     }
-  } while (truncated);
-  hash.Final((byte *)final_etag);
-
-  buf_to_hex((unsigned char *)final_etag, sizeof(final_etag), final_etag_str);
-  snprintf(&final_etag_str[CEPH_CRYPTO_MD5_DIGESTSIZE * 2],  sizeof(final_etag_str) - CEPH_CRYPTO_MD5_DIGESTSIZE * 2,
-           "-%lld", (long long)parts->parts.size());
-  etag = final_etag_str;
-  ldout(s->cct, 10) << "calculated etag: " << final_etag_str << dendl;
-
-  etag_bl.append(final_etag_str, strlen(final_etag_str) + 1);
-
-  attrs[RGW_ATTR_ETAG] = etag_bl;
-
-  if (compressed) {
-    // write compression attribute to full object
-    bufferlist tmp;
-    ::encode(cs_info, tmp);
-    attrs[RGW_ATTR_COMPRESSION] = tmp;
-  }
-
-  target_obj.init(s->bucket, s->object.name);
-  if (versioned_object) {
-    store->gen_rand_obj_instance_name(&target_obj);
-  }
-
-  RGWObjectCtx& obj_ctx = *static_cast<RGWObjectCtx *>(s->obj_ctx);
-
-  obj_ctx.obj.set_atomic(target_obj);
-
-  RGWRados::Object op_target(store, s->bucket_info, *static_cast<RGWObjectCtx *>(s->obj_ctx), target_obj);
-  RGWRados::Object::Write obj_op(&op_target);
-
-  obj_op.meta.manifest = &manifest;
-  obj_op.meta.remove_objs = &remove_objs;
-
-  obj_op.meta.ptag = &s->req_id; /* use req_id as operation tag */
-  obj_op.meta.owner = s->owner.get_id();
-  obj_op.meta.flags = PUT_OBJ_CREATE;
-  op_ret = obj_op.write_meta(ofs, accounted_size, attrs);
-  if (op_ret < 0)
-    return;
+  
+    target_obj.init(s->bucket, s->object.name);
+    if (versioned_object) {
+      store->gen_rand_obj_instance_name(&target_obj);
+    }
+  
+    RGWObjectCtx& obj_ctx = *static_cast<RGWObjectCtx *>(s->obj_ctx);
+  
+    obj_ctx.obj.set_atomic(target_obj);
+  
+    RGWRados::Object op_target(store, s->bucket_info, *static_cast<RGWObjectCtx *>(s->obj_ctx), target_obj);
+    RGWRados::Object::Write obj_op(&op_target);
+  
+    obj_op.meta.manifest = &manifest;
+    obj_op.meta.remove_objs = &remove_objs;
+  
+    obj_op.meta.ptag = &s->req_id; /* use req_id as operation tag */
+    obj_op.meta.owner = s->owner.get_id();
+    obj_op.meta.flags = PUT_OBJ_CREATE;
+    op_ret = obj_op.write_meta(ofs, accounted_size, attrs);
+    if (op_ret < 0)
+      return;
+  
+    // remove the upload obj
+    int r = store->delete_obj(*static_cast<RGWObjectCtx *>(s->obj_ctx),
+  			    s->bucket_info, meta_obj, 0);
+    if (r < 0) {
+      ldout(store->ctx(), 0) << "WARNING: failed to remove object " << meta_obj << dendl;
+    }
+    break;
+  } while(1);
 
-  // remove the upload obj
-  int r = store->delete_obj(*static_cast<RGWObjectCtx *>(s->obj_ctx),
-			    s->bucket_info, meta_obj, 0);
+  int r = l.unlock(&ioctx, raw_meta_oid);
```

Comment:
note that when the call to `delete_obj()` succeeds, this unlock() will fail with `-ENOENT`. it should probably only be called in the `WARNING: failed to remove object` block